### PR TITLE
Add Injected Language Support

### LINKED ISSUE #16: ✨ Injected Languages
Injected languages are important for languages like `HTML` which might embed `CSS` or `JavaScript` code blocks.

## Example Use Cases

- `HTML` implements `CSS`, `JavaScript`
- `JavaScript` implements `regex`, `jsdoc`, `HTML/React`

## Injections.scm

Some grammars that support injections have an additional `injections.scm` file in their `queries` folder.

An example is [tree-sitter-html](https://github.com/tree-sitter/tree-sitter-html):

```
((script_element
  (raw_text) @injection.content)
 (#set! injection.language "javascript"))

((style_element
  (raw_text) @injection.content)
 (#set! injection.language "css"))
```

We need to figure out how to parse and load the associated injected languages in addition to the base language itself.

## Resources
- https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection

### PR DESCRIPTION
**Pull request by thecoolwinter**
# Description

This PR adds support for injected languages using tree-sitter as described in #16 and in the [tree-sitter documentation](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection). Languages can contain _injected_ languages which are described in a language's `injections.scm` file. Some examples of injected languages are:
- HTML contains CSS and Javascript in `style` and `script` tags
- Javascript contains Regex literals
- PHP contains HTML between the `<php` tags
- C++ can contain rawstring literals of arbitrary languages

# Details

This PR is a rework of the `TreeSitterClient` class. Specifically it:
- Adds a `layers` array and `primaryLayer` property. `layers` contains all language layers in the document, and `primaryLayer` is the ID of the document's primary language.
- Each layer is a `LanguageLayer` object. These objects represent an injected language-range(s) combination. Each layer can have one or more range associated with it depending on if it should be [parsed as one document or multiple](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection).
- When editing:
  - Each layer's ranges are updated using a similar algorithm to tree-sitter, and edits are applied to make use of the incremental parsing tree-sitter gives.
  - Each layer is checked for any injections, inserting any new injected layers and keeping track of any 'untouched' layers to remove.
  - Any layers that were not touched are removed.

The highlight query algorithm is largely the same, but keeps track of any ranges not used by any injected layers, and only queries the primary layer for those ranges so as not to override any injected highlights.

# Related Issues

- Closes #16 

# Screenshots

Before, other languages were detected but parsed and highlighted as normal text.
Screenshot 2023-02-28 at 3 08 52 PM

With Injected languages, in this case CSS and JS embedded in HTML and a second layer of Regex embedded in JS embedded in HTML:
Screenshot 2023-03-24 at 2 45 15 PM

### CODE REVIEW

**Comment by lukepistrol** [comment]
Is this still WIP?

**Comment by thecoolwinter** [comment]
> Is this still WIP?

Yeah it is. There are still some glaring issues right now that I thought would be trivial to fix but are ending up being trickier than expected.

File: Sources/CodeEditTextView/Extensions/NSRange+/NSRange+TSRange.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//  Created by Khan Winter on 2/26/23.
+//
+
+import Foundation
+import SwiftTreeSitter
+
+extension NSRange {
+    var tsRange: TSRange {
+        return TSRange(points: .zero..<(.zero),
+                       bytes: (UInt32(self.location) * 2)..<(UInt32(self.location + self.length) * 2))
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/Extensions/NSRange+/NSRange+TSRange.swift:14]: ### [CodeStyle]: 

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
        return TSRange(
            points: .zero..<(.zero),
            bytes: (UInt32(self.location) * 2)..<(UInt32(self.location + self.length) * 2)
        )
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    /// - Parameters:
+    ///   - textView: The text view to use for text.
+    ///   - edit: The edit to act on.
+    ///   - language: The language to use.
+    ///   - readBlock: A callback for fetching blocks of text.
+    /// - Returns: An array of distinct `NSRanges` that need to be re-highlighted.
+    func findChangedByteRanges(textView: HighlighterTextView,
+                               edit: InputEdit,
+                               layer: LanguageLayer,
+                               readBlock: @escaping Parser.ReadBlock) -> [NSRange] {
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift:24]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
    func findChangedByteRanges(
        textView: HighlighterTextView,
        edit: InputEdit,
        layer: LanguageLayer,
        readBlock: @escaping Parser.ReadBlock
    ) -> [NSRange] {
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        let (oldTree, newTree) = calculateNewState(tree: layer.tree,
+                                                   parser: layer.parser,
+                                                   edit: edit,
+                                                   readBlock: readBlock)
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift:28]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
        let (oldTree, newTree) = calculateNewState(
            tree: layer.tree,
            parser: layer.parser,
            edit: edit,
            readBlock: readBlock
        )
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    /// - Parameters:
+    ///   - tree: The tree before an edit used to parse the new tree.
+    ///   - parser: The parser used to parse the new tree.
+    ///   - edit: The edit to apply.
+    ///   - readBlock: The block to use to read text.
+    /// - Returns: (The old state, the new state).
+    internal func calculateNewState(tree: Tree?,
+                                    parser: Parser,
+                                    edit: InputEdit,
+                                    readBlock: @escaping Parser.ReadBlock) -> (Tree?, Tree?) {
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift:53]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
    internal func calculateNewState(
        tree: Tree?,
        parser: Parser,
        edit: InputEdit,
        readBlock: @escaping Parser.ReadBlock
    ) -> (Tree?, Tree?) {
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    ///   - touchedLayers: The set of layers that existed before updating injected layers.
+    ///                    Will have items removed as they are found.
+    ///   - readBlock: A completion block for reading from text storage efficiently.
+    /// - Returns: An index set of any updated indexes.
+    @discardableResult
+    internal func updateInjectedLanguageLayers(textView: HighlighterTextView,
+                                               layer: LanguageLayer,
+                                               layerSet: inout Set<LanguageLayer>,
+                                               touchedLayers: inout Set<LanguageLayer>,
+                                               readBlock: @escaping Parser.ReadBlock) -> IndexSet {
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Edit.swift:103]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
    internal func updateInjectedLanguageLayers(
        textView: HighlighterTextView,
        layer: LanguageLayer,
        layerSet: inout Set<LanguageLayer>,
        touchedLayers: inout Set<LanguageLayer>,
        readBlock: @escaping Parser.ReadBlock
    ) -> IndexSet {
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Highlight.swift [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    /// Queries the given language layer for any highlights.
+    /// - Parameters:
+    ///   - layer: The layer to query.
+    ///   - textView: A text view to use for contextual data.
+    ///   - range: The range to query for.
+    /// - Returns: Any ranges to highlight.
+    internal func queryLayerHighlights(layer: LanguageLayer,
+                                       textView: HighlighterTextView,
+                                       range: NSRange) -> [HighlightRange] {
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient+Highlight.swift:22]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
    internal func queryLayerHighlights(
        layer: LanguageLayer,
        textView: HighlighterTextView,
        range: NSRange
    ) -> [HighlightRange] {
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient+LanguageLayer.swift [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        ///   - supportsInjections: Set to true when the langauge supports the `injections` query.
+        ///   - tree: The tree-sitter tree generated while editing/parsing a document.
+        ///   - languageQuery: The language query used for fetching the associated `queries.scm` file
+        ///   - ranges: All ranges this layer acts on. Must be kept in order and w/o overlap.
+        init(id: TreeSitterLanguage,
+             parser: Parser,
+             supportsInjections: Bool,
+             tree: Tree? = nil,
+             languageQuery: Query? = nil,
+             ranges: [NSRange]) {
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient+LanguageLayer.swift:27]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
        init(
            id: TreeSitterLanguage,
            parser: Parser,
            supportsInjections: Bool,
            tree: Tree? = nil,
            languageQuery: Query? = nil,
            ranges: [NSRange]
        ) {
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient.swift [hunk 8]
```diff
@@ -18,46 +18,85 @@ import SwiftTreeSitter
...
+        layers.removeAll()
+
+        primaryLayer = codeLanguage.id
+        layers = [
+            LanguageLayer(id: codeLanguage.id,
+                          parser: Parser(),
+                          supportsInjections: codeLanguage.additionalHighlights?.contains("injections") ?? false,
+                          tree: nil,
+                          languageQuery: TreeSitterModel.shared.query(for: codeLanguage.id),
+                          ranges: [])
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient.swift:70]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
            LanguageLayer(
                id: codeLanguage.id,
                parser: Parser(),
                supportsInjections: codeLanguage.additionalHighlights?.contains("injections") ?? false,
                tree: nil,
                languageQuery: TreeSitterModel.shared.query(for: codeLanguage.id),
                ranges: []
            )
```

File: Sources/CodeEditTextView/TreeSitter/TreeSitterClient.swift [hunk 9]
```diff
@@ -75,143 +114,164 @@ final class TreeSitterClient: HighlightProviding {
...
-                                 range: NSRange,
-                                 completion: @escaping (([HighlightRange]) -> Void)) {
-        guard let rootNode = tree.rootNode else {
-            completion([])
-            return
+        // Query primary for any ranges that weren't used in the injected layers.
+        for range in injectedSet.rangeView {
+            highlights.append(contentsOf: queryLayerHighlights(layer: layers[0],
+                                                               textView: textView,
+                                                               range: NSRange(range)))
```
Comment (lukepistrol) [inline Sources/CodeEditTextView/TreeSitter/TreeSitterClient.swift:215]: ### [CodeStyle]:

Maybe try to not indent to much. In the `CodeEdit` repo this is not allowed by SwiftLint. We should keep it consistent.

```suggestion
            highlights.append(contentsOf: queryLayerHighlights(
                layer: layers[0],
                textView: textView,
                range: NSRange(range)
            ))
```

**Review by lukepistrol — CHANGES_REQUESTED** [review]
Just a couple of code style considerations about function parameter indentation. I provided fixes for most of them.

Refs: [Code Style](https://github.com/CodeEditApp/CodeEdit/wiki/Code-Style#parameter-alignment)

File: Sources/CodeEditTextView/Extensions/NSRange+/NSRange+Comparable.swift [hunk 10]
```diff
@@ -0,0 +1,18 @@
+//
+//  File.swift
```
Comment (0xWDG) [inline Sources/CodeEditTextView/Extensions/NSRange+/NSRange+Comparable.swift:2]: Maybe update the file name comment

**Review by 0xWDG — DISMISSED** [review]

**Review by lukepistrol — APPROVED** [review]
LGTM 👍

**Review by austincondiff — APPROVED** [review]
This is really great work @thecoolwinter! 👏

**Comment by lukepistrol** [comment]
@thecoolwinter Are there any breaking changes regarding the public API? Need to know when releasing a new version.

**Comment by thecoolwinter** [comment]
> @thecoolwinter Are there any breaking changes regarding the public API? Need to know when releasing a new version.

Yes, there's an additional method on `HighlightProvider` which external highlight providers will have to implement when updating.

**Pull request merged**